test.js

```diff
@@ -1,53 +1,136 @@
+	inputStream.destroy();
+	await once(stream, 'close');
+	t.false(stream.readableEnded);
+	t.is(stream.errored, null);
+	t.true(stream.readableAborted);
+	t.true(stream.closed);
+	t.true(stream.destroyed);
 });
 
 test('handles no input', async t => {
-	const result = await mergeStreams([]).toArray();
-	t.deepEqual(result, []);
+	const stream = mergeStreams([]);
+	t.deepEqual(await stream.toArray(), []);
 });
+
+test('validates argument is an array', t => {
+	t.throws(() => {
+		mergeStreams(Readable.from([]));
+	}, {message: /Expected an array/});
+});
+
+test('validates arguments are streams', t => {
+	t.throws(() => {
+		mergeStreams([false]);
+	}, {message: /Expected a readable stream/});
+});
+
+const testObjectMode = async (t, firstObjectMode, secondObjectMode, mergeObjectMode) => {
+	const stream = mergeStreams([
+		Readable.from(['a'], {objectMode: firstObjectMode}),
+		Readable.from(['b'], {objectMode: secondObjectMode}),
+	]);
+	t.is(stream.readableObjectMode, mergeObjectMode);
+	await stream.toArray();
+};
+
+test('is not in objectMode if no input stream is', testObjectMode, false, false, false);
+test('is in objectMode if only some input streams are', testObjectMode, false, true, true);
+test('is in objectMode if all input streams are', testObjectMode, true, true, true);
+
+const testHighWaterMarkAmount = async (t, firstObjectMode, secondObjectMode, highWaterMark) => {
+	const stream = mergeStreams([
+		Readable.from(['a', 'b'], {highWaterMark: 4, objectMode: firstObjectMode}),
+		Readable.from(['c', 'd'], {highWaterMark: 2, objectMode: secondObjectMode}),
+	]);
+	t.is(stream.readableHighWaterMark, highWaterMark);
+	t.is(stream.writableHighWaterMark, 0);
+	await stream.toArray();
+};
+
+test('highWaterMark is the maximum of non-object input streams', testHighWaterMarkAmount, false, false, 4);
+test('highWaterMark is the maximum of object input streams', testHighWaterMarkAmount, true, true, 4);
+test('highWaterMark is the maximum of object streams if mixed with non-object ones', testHighWaterMarkAmount, false, true, 2);
+
+const testBufferSize = async (t, objectMode) => {
+	const highWaterMark = getDefaultHighWaterMark(objectMode);
+	const oneStream = new PassThrough({highWaterMark});
+	const twoStream = new PassThrough({highWaterMark});
+	const stream = mergeStreams([oneStream, twoStream]);
+
+	// Each PassThrough has a read + write buffer, except the merged stream which has no write buffer
+	// Therefore, there are 5 buffers of size `highWaterMark`
+	const bufferCount = 5;
+
+	let writeCount = 0;
+	while (oneStream.write('.') && twoStream.write('.')) {
+		writeCount += 1;
+		// eslint-disable-next-line no-await-in-loop
+		await scheduler.yield();
+	}
+
+	t.is(writeCount, Math.ceil((highWaterMark - 1) * bufferCount / 2));
+
+	let readCount = 0;
+	while (stream.read() !== null) {
+		readCount += 1;
+		// eslint-disable-next-line no-await-in-loop
+		await scheduler.yield();
+	}
+
+	t.is(readCount, bufferCount);
+
+	t.pass();
+};
```

Comment:
Each `PassThrough` has one read buffer + one write buffer of size `highWaterMark`. The merged stream too, except it does not have a write buffer (`writableHighWaterMark` is `0`).

Therefore, the streams will be full after writing `5 * highWaterMark` bytes.
Then, the merged stream will consume all those bytes in exactly 5 reads, returning each `highWaterMark` bytes.